pulser/waveforms.py

```diff
@@ -524,11 +441,14 @@ def __repr__(self) -> str:
     def __getitem__(self, index_or_slice:
                     Union[int, slice]) -> Union[float, np.ndarray]:
         if isinstance(index_or_slice, slice):
-            self._check_slice(index_or_slice)
-            return cast(np.ndarray, self._samples[index_or_slice])
+            return super().__getitem__(index_or_slice)
         else:
-            self._check_index(index_or_slice)
-            return cast(float, self._samples[index_or_slice])
+            index: int = self._check_index(index_or_slice)
+            if index == 0:
+                return self._start
+            if index == self._duration - 1:
+                return self._stop
+            return super().__getitem__(index)
```

Comment:
I know I agreed to leaving these cases in, but could please you check if there's actually any significant performance gain vs. just getting the value from `self._samples`?

Reply:
Checking indexes first is indeed faster, but not so much (less than twice):

```
Benchmarking access to the first value (1 million times):
* Direct access    : 0.09557080000000018
* Check index first: 0.05514600000000014

Benchmarking access to the last value (1 million times):
* Direct access    : 0.09351679999999996
* Check index first: 0.06454629999999995
```

But the absolute difference in duration is extremely tiny (the difference between measurements above, in seconds, divided by 1 million...).

I think we should remove the checks, considering:
* This does not really affect performance
* It makes code more verbose
* The calls on the first and alst value probably don't occur really often

Reply:
I agree. Isn't this also the case for the other special cases, though?

Reply:
Yes, yes, I meant removing the checks for all the special cases. Let's avoid unnecessary complexity whenever performance gains are not that important.